internal/command/user_grant.go

```diff
@@ -35,11 +38,14 @@ func (c *Commands) AddUserGrant(ctx context.Context, usergrant *domain.UserGrant
 	return userGrantWriteModelToUserGrant(addedUserGrant), nil
 }
 
-func (c *Commands) addUserGrant(ctx context.Context, userGrant *domain.UserGrant, resourceOwner string) (command eventstore.Command, _ *UserGrantWriteModel, err error) {
+func (c *Commands) addUserGrant(ctx context.Context, userGrant *domain.UserGrant, resourceOwner *string, check UserGrantPermissionCheck) (command eventstore.Command, _ *UserGrantWriteModel, err error) {
 	if !userGrant.IsValid() {
 		return nil, nil, zerrors.ThrowInvalidArgument(nil, "COMMAND-kVfMa", "Errors.UserGrant.Invalid")
 	}
-	err = c.checkUserGrantPreCondition(ctx, userGrant, resourceOwner)
+	err = c.checkUserGrantPreCondition(ctx, userGrant, resourceOwner, check)
```

Comment:
You could remove the resourceOwner as parameter as the userGrant can contain a resourceOwner already, then return the resourceOwner from the preCondition and check if the resourceOwner provided and either of the project or project grant are equal.

Reply:
But not of the project grant.
It's really about resolving this https://github.com/zitadel/zitadel/pull/9978#discussion_r2170953671 or am I missing something?